stew/conio.nim

```diff
@@ -0,0 +1,337 @@
+        if int16(buffer[^1]) == int16(0x0A):
+          if len(buffer) > 1:
+            if int16(buffer[^2]) == int16(0x0D):
+              buffer.setLen(len(buffer) - 2)
+            else:
+              buffer.setLen(len(buffer) - 1)
+          else:
+            buffer.setLen(len(buffer) - 1)
+        elif int16(buffer[^1]) == int16(0x0D):
+          buffer.setLen(len(buffer) - 1)
+
+      # Convert Windows UTF-16 encoded string to UTF-8 encoded string.
+      if len(buffer) > 0:
+        var pwd = ""
+        let bytesNeeded = wideCharToMultiByte(CP_UTF8, 0'u32, addr buffer[0],
+                                              cint(len(buffer)), nil,
+                                              cint(0), nil, nil)
+        if bytesNeeded <= cint(0):
+          return err(ioLastError())
+        pwd.setLen(bytesNeeded)
+        let cres = wideCharToMultiByte(CP_UTF8, 0'u32, addr buffer[0],
+                                       cint(len(buffer)), addr pwd[0],
+                                       cint(len(pwd)), nil, nil)
+        if cres == cint(0):
+          return err(ioLastError())
+        ok(pwd)
+      else:
+        ok("")
+
+  proc writeConsoleOutput(data: string): IoResult[void] =
+    if len(data) == 0:
+      return ok()
+
+    let hConsoleOutput =
+      block:
+        let res = getStdHandle(STD_OUTPUT_HANDLE)
+        if res == INVALID_HANDLE_VALUE:
+          return err(ioLastError())
+        res
+
+    let prevOutputCP =
+      block:
+        let res = getConsoleOutputCP()
+        if res == cuint(0):
+          return err(ioLastError())
+        res
+
+    if isConsoleRedirected(hConsoleOutput):
+      # If STDOUT is redirected we should use WriteFile() because WriteConsole()
+      # is not working for such types of STDOUT.
+      if setConsoleOutputCP(CP_UTF8) == 0'i32:
+        return err(ioLastError())
+
+      let bytesToWrite = uint32(len(data))
+      var bytesWritten: uint32
+      let wres = writeFile(hConsoleOutput, cast[pointer](unsafeAddr data[0]),
+                           bytesToWrite, bytesWritten, nil)
+      if wres == 0'i32:
+        let errCode = ioLastError()
+        discard setConsoleOutputCP(prevOutputCP)
+        return err(errCode)
+
+      if setConsoleOutputCP(prevOutputCP) == 0'i32:
+        return err(ioLastError())
+    else:
+      if setConsoleOutputCP(CP_UTF8) == 0'i32:
+        return err(ioLastError())
+
+      let widePrompt = newWideCString(data)
+      var charsWritten: uint32
+      let wres = writeConsole(hConsoleOutput, cast[pointer](widePrompt),
+                              uint32(len(widePrompt)), charsWritten, nil)
+      if wres == 0'i32:
+        let errCode = ioLastError()
+        discard setConsoleOutputCP(prevOutputCP)
+        return err(errCode)
+
+      if setConsoleOutputCP(prevOutputCP) == 0'i32:
+        return err(ioLastError())
+    ok()
+
+elif defined(posix):
+  import posix, termios
+
+  proc isConsoleRedirected(consoleFd: cint): bool =
```

Comment:
In Chronicles, I too need to check if `stdin` is redirected. Let's export something that would allow me perform this check in cross-platform way.